src/lasdi/latent_space.py

```diff
@@ -1,175 +1,482 @@
+
+        The image of x under the network's layers. If self.reshape_index == -1 and 
+        self.reshape_shape has k elements, then we reshape the output so that the final k elements
+        of its shape match those of self.reshape_shape.
+        """
+
+        # If the reshape_index is 0, we need to reshape x before passing it through the first 
+        # layer.
         if (self.reshape_index == 0):
-            # make sure the input has a proper shape
+            # Make sure the input has a proper shape. There is a lot going on in this line; let's
+            # break it down. If reshape_index == 0, then we need to reshape the input, x, before
+            # passing it through the layers. Let's assume that reshape_shape has k elements. Then,
+            # we need to squeeze the final k dimensions of the input, x, so that the resulting 
+            # tensor has a final dimension size that matches the input dimension size for the first
+            # layer. The check below makes sure that the final k dimensions of the input, x, match
+            # the stored reshape_shape.
             assert(list(x.shape[-len(self.reshape_shape):]) == self.reshape_shape)
-            # we use torch.Tensor.view instead of torch.Tensor.reshape,
-            # in order to avoid data copying.
+            
+            # Now that we know the final k dimensions of x have the correct shape, let's squeeze 
+            # them into 1 dimension (so that we can pass the squeezed tensor through the first 
+            # layer). To do this, we reshape x by keeping all but the last k dimensions of x, and 
+            # replacing the last k with a single dimension whose size matches the dimensionality of
+            # the domain of the first layer. Note that we use torch.Tensor.view instead of 
+            # torch.Tensor.reshape in order to avoid data copying.
             x = x.view(list(x.shape[:-len(self.reshape_shape)]) + [self.layer_sizes[self.reshape_index]])
 
-        for i in range(self.n_layers-2):
-            x = self.fcs[i](x) # apply linear layer
-            if (self.use_multihead):
-                x = self.apply_attention(self, x, i)
-            else:
-                x = self.act(x)
+        # Pass x through the network layers (except for the final one, which has no activation 
+        # function).
+        for i in range(self.n_layers - 1):
+            x : torch.Tensor = self.layers[i](x)   # apply linear layer
+            x : torch.Tensor = self.act(x)      # apply activation
 
-        x = self.fcs[-1](x)
+        # Apply the final (output) layer.
+        x = self.layers[-1](x)
 
+        # If the reshape_index is -1, then we need to reshape the output before returning. 
         if (self.reshape_index == -1):
-            # we use torch.Tensor.view instead of torch.Tensor.reshape,
-            # in order to avoid data copying.
+            # In this case, we need to split the last dimension of x, the output of the final
```

Comment:
Maybe use key words only?
```
split the last dimension of x, the output of the final layer, to match the reshape_shape.
Note that we use torch.Tensor.view instead of torch.Tensor.reshape in order to avoid data copying.
```